### ISSUE TITLE
Slow-reader problem

### ISSUE THREAD

**Issue by enobufs** [pion/sctp#40]
## The issue
It appears that SCTP is reading incoming data slow in a high bandwidth usage, causing a buffer (packetio) at ICE layer being filled up with a lot of incoming data (cause a lot of latency before reaching SCTP layer) - pion/ice#12.

(Excerpt from the discussion in the PR #39)
[syscall.pdf](https://github.com/pion/sctp/files/3087622/syscall.pdf)
Looking at the trace output (go tool trace), it appears, what association's readLoop is blocked by is not the mutex of stream layer, but the syscall "sendto". (underlying (UDP) socket is a blocking socket)
To solve the "slow-reader" problem, we'd probably need a drastic change in SCTP, such as stop sending data (a reply) on handleInbound(), and use another goroutine for the immediate replies (control packets), which would raise another issue - how much can sctp layer buffer those immediate replies?

Related issue: pion/ice#12 and #32

## Approach (proposal)
* Introduce another goroutine, `writeLoop` in association layer
* WriteSCTP() stores data to pending data, wake up the `writeLoop` and return immediately.
* handleInbound() (the readLoop routine) DO NOT CALL association.send(). Instead, it should schedule the transmission of immediate control packets for `writeLoop` to send it, and exit the routine as soon as possible (similar to delayed-ack, but immediately)
* Also consider avoiding read and write routines waiting on the same mutex at the stream layer. (introduce a mutex inside reassembly queue, etc.

TODO: evaluate if the above proposal is viable first.

**Comment by backkem** [comment]
What about making WriteSCTP blocking once we run too far ahead (to much in flight)? That way we just slow down the entire pipeline instead of flooding the remote or the pending data buffer.

**Comment by enobufs** [comment]
How fast the outgoing data goes out onto wire depends on the congestion window now, so you cannot fully control the outgoing traffic by blocking WriteSCTP.

With the delayed-ack, about only 1% of handleInboud() calls are calling a.send() (to send immediate SACK or FowardTSN) but that's when there was no packet loss. Once packet loss occurs, every single handInbound() call will call a.send(), until all lost packets are recovered, that would slow down the reading speed because a.send() would block when sending data at a high rate.

Healthy "back pressure" should happen with a receipt of SACK indicating there was a packet loss in timely manner, so that cwnd would be reduced to a half (but will be recovered with fast retransmission). But if handleInbound() is not processing fast enough, it creates a long backlog in the ICE buffer, SACKs (that would inform of the loss) are coming back to too late for the sender to reduce the cwnd in time.  Also, receiver know much data it (sctp) can hold on to (the rwnd), and would discard inbound packet beyond rwnd, which would also cause the **healthy** back pressure too, but the data is actually buffered at ICE level, sctp layer does not even know about the arrival of those packets...  that is, sctp congestion mechanism isn't functioning.

This is what I think is going on...  So, I believe we need to make the readLoop fast enough not to cause the large backlog in the ICE buffer.

My TODO: try to reduce ICE buffer size (from 1 MB to 64KB or so)  to cause packet loss earlier. it might mitigate the situation...

**Comment by enobufs** [comment]
@backkem, I found a bug in fast retransmission, not recovering well from contiguous losses. (just wanted to let you know I am working on it)
